File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/DropCardOnDiscardPileCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.decks.Card
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class DropCardOnDiscardPileCmd(
                                  gameIdentity: GameIdentity,
                                  playerIdentity: PlayerIdentity,
                                  card: Card
                                ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class DropCardOnDiscardPileHandler(
                                    gameRepositoryPort: GameRepositoryPort
                                  )
                                  (implicit val ec: scala.concurrent.ExecutionContext)
  extends CommandHandler[DropCardOnDiscardPileCmd] {

  override def handleAsync(command: DropCardOnDiscardPileCmd): Future[Unit] = {
    for {
      gameExecution <- gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity)
      updatedGameExecution = gameExecution.dropCardOnDiscardPile(playerIdentity = command.playerIdentity, card = command.card)
    } yield gameRepositoryPort.save(updatedGameExecution)
  }

  override def handle(command: DropCardOnDiscardPileCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val gameExecution = Await.result(gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity), 500 millis)
    val updatedGameExecution = gameExecution.dropCardOnDiscardPile(playerIdentity = command.playerIdentity, card = command.card)
    gameRepositoryPort.save(updatedGameExecution)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commandModel/commands/DropCardOnDiscardPileCmd.kt
package com.abaddon83.burracoGame.commandModel.commands

import com.abaddon83.burracoGame.commandModel.models.burracoGameExecutions.BurracoGameExecutionTurnExecution
import com.abaddon83.burracoGame.readModel.ports.BurracoGameReadModelRepositoryPort
import com.abaddon83.burracoGame.commandModel.models.decks.Card
import com.abaddon83.burracoGame.commandModel.models.games.GameIdentity
import com.abaddon83.burracoGame.commandModel.models.players.PlayerIdentity
import com.abaddon83.burracoGame.commandModel.ports.BurracoGameRepositoryPort
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import com.abaddon83.utils.logs.WithLog
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class DropCardOnDiscardPileCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        val card: Card,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity, playerIdentity: PlayerIdentity, card: Card): this(
            gameIdentity = gameIdentity,
            playerIdentity = playerIdentity,
            card = card,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis()))
}

class DropCardOnDiscardPileHandler() : CommandHandler<DropCardOnDiscardPileCmd>, KoinComponent, WithLog() {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(command: DropCardOnDiscardPileCmd) {
        executeCmd(command)
    }

    override suspend fun handleAsync(command: DropCardOnDiscardPileCmd) {
        executeCmd(command)
    }

    private fun executeCmd(command: DropCardOnDiscardPileCmd) {
        when (val game = repository.getById(command.gameIdentity)) {
            is BurracoGameExecutionTurnExecution -> repository.save(game
                    .dropCardOnDiscardPile(playerIdentity = command.playerIdentity, card = command.card))
            else -> warnMsg("GameIdentity ${command.gameIdentity} not found exist")
        }
    }
}